base/strings/search.jl

```diff
@@ -308,30 +292,18 @@ function _rsearchindex(s::AbstractString,
         return 1 <= i <= nextind(s, lastindex(s)) ? i :
                throw(BoundsError(s, i))
     end
-    t = t isa AbstractString ? reverse(t) : t
-    rs = reverse(s)
-    l = lastindex(s)
-    t1, j2 = next(t, start(t))
+    t1, trest = Iterators.peel(Iterators.reverse(t))
     while true
         i = findprev(equalto(t1), s, i)
         i === nothing && return 0
-        c, ii = next(rs, reverseind(rs, i))
-        j = j2; k = ii
-        matched = true
-        while !done(t, j)
-            if done(rs, k)
-                matched = false
-                break
-            end
-            c, k = next(rs, k)
-            d, j = next(t, j)
-            if c != d
-                matched = false
-                break
-            end
+        a = Iterators.Stateful(trest)
+        ii = prevind(s, i)
+        matched = all(zip(Iterators.reverse(pairs(s[1:ii])), a)) do ((idx, c), d)
+            i = idx
```

Comment:
The compiler can't optimize this (a do block closure). Also hard to read.

Reply:
fair enough, I'll change it to explicitly take the next iteration and compute nextind.